packages/contracts-bedrock/contracts/test/OptimismPortal.t.sol

```diff
@@ -310,6 +261,242 @@ contract OptimismPortal_Test is Portal_Initializer {
+            latestBlockhash: bytes32(0)
+        });
+        vm.mockCall(
+            address(op.L2_ORACLE()),
+            abi.encodeWithSelector(L2OutputOracle.getL2Output.selector),
+            abi.encode(Hashing.hashOutputRootProof(outputRootProof), _proposedBlockNumber)
+        );
+        vm.expectRevert("OptimismPortal: insufficient gas to finalize withdrawal");
+        op.finalizeWithdrawalTransaction{ gas: gasLimit }(
+            insufficientGasTx,
+            _proposedBlockNumber,
+            outputRootProof,
+            withdrawalProof
+        );
+    }
+
+    // Test: finalizeWithdrawalTransaction reverts if the proof is invalid due to non-existence of
+    // the withdrawal.
+    function test_finalizeWithdrawalTransaction_revertsOninvalidWithdrawalProof() external {
+        // modify the default test values to invalidate the proof.
+        _defaultTx.data = hex"abcd";
+        vm.expectRevert("OptimismPortal: invalid withdrawal inclusion proof");
+        op.finalizeWithdrawalTransaction(
+            _defaultTx,
+            _proposedBlockNumber,
+            _outputRootProof,
+            _withdrawalProof
+        );
+    }
+
+    // Utility function used in the subsequent test. This is necessary to assert that the
+    // reentrant call will revert.
+    function callAndExpectRevert(bytes calldata data) external {
+        vm.expectRevert("OptimismPortal: can only trigger one withdrawal per transaction");
+        // Arguments here don't matter, as the require check is the first thing that happens.
+        op.finalizeWithdrawalTransaction(_defaultTx, 0, _outputRootProof, hex"");
+    }
+
+    // Test: finalizeWithdrawalTransaction reverts if a sub-call attempts to finalize another
+    // withdrawal.
+    function test_finalizeWithdrawalTransaction_revertsOnReentrancy() external {
+        // Setup the Oracle to return an output with a finalized timestamp
+        vm.warp(op.FINALIZATION_PERIOD_SECONDS() + 1000);
+        uint256 finalizedTimestamp = block.timestamp - op.FINALIZATION_PERIOD_SECONDS() - 1;
+        bytes memory data = abi.encodeWithSelector(this.callAndExpectRevert.selector);
+        (
+            bytes32 stateRoot,
+            bytes32 storageRoot,
+            bytes32 outputRoot,
+            bytes32 withdrawalHash,
+            bytes memory withdrawalProof
+        ) = ffi.getFinalizeWithdrawalTransactionInputs(_defaultTx);
+        // ) = getFinalizeWithdrawalTransactionInputs(_n, _s, address(this), 0, _g, data);
+        Types.OutputRootProof memory outputRootProof = Types.OutputRootProof({
+            version: bytes32(0),
+            stateRoot: stateRoot,
+            withdrawerStorageRoot: storageRoot,
+            latestBlockhash: bytes32(0)
+        });
+        vm.mockCall(
+            address(op.L2_ORACLE()),
+            abi.encodeWithSelector(L2OutputOracle.getL2Output.selector),
+            abi.encode(Types.OutputProposal(outputRoot, finalizedTimestamp))
+        );
+
```

Comment:
Could you add a `vm.expectCall` here? Its not clear what the assertion is that is being made just by reading the test

Reply:
Yeah, this test is pretty hard to follow (and was actually broken before). 
I've added several checks with `vm.expectX()` and `assert()`, as well as comments to explain what's going on.